### ISSUE TITLE
Prioritization of the leaf crates stabilization and dependency graph

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#2373]
## Summary

Let's focus on stabilizing the most important crates: `units`, `primitives`, `keys`, `address`, in this order.

## Motivation

The above mentioned crates provide fundamental types used in inter-crate communication. By being unstable a breaking change in our crate is a much bigger annoyance then it'd be otherwise: the whole ecosystem breaks and everyone is waiting on someone else to upgrade. This can take many months. Therefore it should be our priority to get them ready ASAP.

Further, these crates have most of the API in the state where I'm confident could be stabilized. Being used almost everywhere, their API is also battle-tested, so its stability is not just a theory.

## Design constraints

A huge care needs to be taken when doing this as we could cause accidental breakage or paint ourselves into a corner unable to develop the downstream crates. I've identified these constraints:

* No crate may publicly depend on an unstable crate (obviously)
* To make development possible we need a `bitcoin-unstable` cfg; as discussed it needs to be a cfg, not feature
* The downstream crates should not require the `bitcoin-unstable` cfg to implement any of the existing functionality (would be way too annoying otherwise)
* Do not duplicate non-trivial code, doing so often leads to horrible problems
* Prefer not to create more micro-crates if possible (containing one trait or so)

## Challenges

The constraints lead to some difficulties around consensus encoding and hashes but I think I figured them out.

One issue is that `Transaction` needs serialization code to implement `compute_txid`/`compute_wtxid` however the `push_decode` part is highly unstable and we don't want to stabilize the existing traits either. This can be solved by putting `AsRawTransaction` similar to `AsTransaction` to the `internals` crate which also implements transaction encoding over this trait. But rather than implementing `AsRawTransaction` for `Transaction` it'll be implemented over a private wrapper.

Another issue is that the ideal `Hash` trait definition requires advanced Rust features and we don't want to wait for super-high MSRV. I think that we can actually get rid of the trait and just use inherent items. The API will be less annoying anyway. This means we need to stabilize *the output* of the `hash_newtype!` macro. I think we can realistically do that once it supports inherent items.

Public keys need to be parsed for that we need `secp256k1`. We don't need it to be stable if we don't expose the internal key but it must have non-context API for critical operations.

## Proposed dependency graph

(Only parts relevant to this issue.)

* Everything depends on internals (non-optionally)
* `internals` optionally depends on `push_decode` (activated for `primitives` and `consensus-encoding`)
* `primitives` depend on `units` and `hashes`
* `consensus_encoding` depends on `primitives`
* `address` depends on `keys` and optionally on `primitives`

**Comment by apoelstra** [comment]
Agree with the prioritization and high-level observations. Can you say more about:

* What `AsRawTransaction` would do? (Maybe this is not the issue for it.)
* What features we'd want from `Hash` trait ideally. Somewhere I recall we came to a rough consensus that our 1.0 MSRV would be 1.6x or thereabouts.

**Comment by Kixunil** [comment]
> What `AsRawTransaction` would do?

Effectively a copy of `AsTransaction` from #2347 with only modification being use of raw types (integers, byte slices) to resolve the dependency loop. It'd be considered an implementation detail and only stay there as long as other crates are unstable. Once stable, the implementation would change and the trait would get removed.

> * What features we'd want from `Hash` trait ideally. Somewhere I recall we came to a rough consensus that our 1.0 MSRV would be 1.6x or thereabouts.

Allow writing `fn to_byte_array(self) -> [u8; Self::LEN]`, though IIRC it may require the feature that is not even in stable and is somewhat broken in nightly too. (I might be confusing this with something else.) But from what I can observe so far: both GAT and weak dependencies will be a must when fully stable, so 1.65 IIRC. But the leaf crates don't need this. Actually that's what I wanted to mention above and forgot.

**Comment by apoelstra** [comment]
I think we should try the `to_byte_array` thing. It doesn't look like it needs GATs, just regular ATs, which have been around for a long time.

The `AsRawTransaction` thing makes sense, thanks. Though I wonder if we want to make this a permanent part of the API (and have a corresponding `Raw` type for some other types, notably `Block` and `BlockHeader`). It seems like it may be useful in some obscure low-level usecases.

OTOH it would be easy to add (or expose) post-1.0 so this isn't a decision we need to make now.

**Comment by Kixunil** [comment]
I don't see when it'd be useful. `Script` is already freely convertible, so only `Witness` is a bit obscure type. We could do something about that too. I was already thinking of having unsized `Witness` for zero-copy deserialization however it'd loose the index which is non issue when serializing but can be a problem when accessing.
